# VCF Parsing in R

## Step 1: Function to parse for individual vcf file
1. Read in vcf file
2. Extract DP, AD
3. Output histogram of raw depth before Filtering
4. Filter data for only sites with read depth greater than 20
3. Separate AD in Allele 1 and Allele 2
4. Make column for minor allele
5. Make column for sum of alleles
6. Make column for minor allele frequency - output histogram
7. Make column for third allele frequency (DP-SumAD/DP) - output histogram
8. Filter by minor allele frequency <0.10 and third allele frequency <0.01
9. Make column with locus name - option to save this table here
10. Make new data frame averaging minor allele freq and depth by overall locus
11. Plot by gene frequencies/heterozygosity likelihood

## Step 2: Loop function over multiples
1. Run VCF tables for 10-15 more specimens to get 20 total and an idea of what the histograms look like

## Future Steps
Determine if need locus or site names to exclude from larger analysis
Determine which specimens to fully exclude
Run for all specimens

# 11/16/21 Notes
- Mpileup assumes diploid, need to figure out how to remove this constraint
  - run a handful of genes with ploidy not set and then ploidy 3

- Use alignment from Luke to make raxml tree for COI
  - COI flag to check for labeling errors

- take a single species from Calder's data and run it through to see allelic read occurrence
  - email calder and ask about one good (not contam) spp and need raw data and assembly
