#2021-01-12 Meeting with Chenhong and Hao

	- Chenhong working on shad genomes (tropical and temperate) and chinese perch
	- could pull loci from pseudoliparis swieri genome
	- widespread contam in dataset
	- struggling to tell if loci similar because contaminated or loci conserved
	- hoping to look at hypervariable loci
	- hoping the heterozygosity will be able to tell us something about contamination
	- two individual from
	- assemble the gene and use consensus to call snps
	- how to check how many contigs per locus
		- pull out best 2 contigs per locus
		- heterozygosity lost at step where we lose assembler like sga (they only pic the consensus or single allele)
		- easiest way is to map them an call the snps
		- next gen sequencing reads are too short so no way to do phasing unless you have long reads like nanopore
	- detect contamination script defined by Hao
		- define probabilty that they are closely related
		- compare p-distance among all samples
		- use structure to see if really closel
		- default value is 0.98
		- if two sequences are close and above 0.98 then potentially contaminated
		- check percentage of contaminated sequences among all the sequences
- compare contigs and maybe even reads between two samples
	- need to map reads to reference first and then check the similarities in columns (reads are 150 bps)
	- extract camers from raw reads and look at one against the other pair by pair to id the problematic samples
	- akin to running detect
	- script exists for inline but script not published/on github
	- Chenhong's lab doing lots of steps to reduce bench contamination
	- for assembly - look at results from assembly and map the reads
	- map reads to contigs and see what percent of reads map to contigs
	- Note from Luke: see if we can pull out snp data for heterozygotes
	 	- mine the snps from the raw reads
		- use reference, map fastq and call snps, can do lots of things with snp dataset
	- Hao doing test with camers to detect contamination
	- If lot of contaminant in sample then heterozygosity will show that (3 or 4 alleles at given snp)
	- Hao advice for snp calling: after calling, filter out those with weird frequencies. Normal frequencies should be 0.5 or 1. Like 0.3 0.2 are weird. 
