What is a heterozyogous sites

filter 1 - read depth > 20 to start
filter 2 - get actual minor allele by adding column get rid of minor allele frequencies < 0.10
filter 3 - make locus column (rename everything before the underscore in rownames)

new data frame with new locus name and average minor allele frequency and depth by overall locus
use this data frame on a species by species basis, what is heterozygosity looking likelihood

Run VCF tables do this for 10-15 to get up to like 20 total an idea of what the histograms look like


(DP-Totallele)/DP = third allele frequency - compare to Illumina sequencing error (less than 1% according to website)


Histogram of raw data depth
Then filter by depth
Minor  allele and third allele freqs - histograms
filter by freqs
Rename by locus and get avg - plot by gene frequencies 


311 questions:
