---
title: "Ch2Results_Figures"
author: "Jennifer Gardner"
date: "1/24/2022"
output: html_document
---
# Figures to make
  1. All the trees
  2. Histogram of identity
  3. VCF histograms


load dependencies
```{r}
library(tidyverse)
library(ape)
library(rstatix)
library(phytools)
library(mgcv)
library(truncnorm)
library(vcfR)
library(stringr)
source("../12_bowtie/vcf-parse-function.R")
```


## Percent Identical Histogram
Load in data and standardize matrix
```{r}
identical_sequences <- read.csv("Identical_Sequences_files/Identical_Sequences.csv", row.names = 1)
total_sequences <- read.csv("Identical_Sequences_files/Total_Sequences.csv", row.names = 1)
percent_identical <- identical_sequences/total_sequences
```

Compare percent identical between members of same species (same_spp_pairs.csv; n =35 spp pairs) vs.
Percent identical between members of same clade but no same species (same_clade_pairs.csv; 162 pairs) vs.
percent identical between all spp pairs not in the same Orr et al. small clades (total.pairs - same_clade_spp_pairs.csv)
  Note: makeup of pairs done in percent_ident_groups.xlsx
  
Get list of all pairwise comparisons (should be 4465 for 95 pairwise comparisons; make from all_lip_samples.csv)
```{r}
all.lip.samples <- read.csv("Identical_Sequences_files/all_lip_samples.csv", header = FALSE)
all.lip.samples <- as.vector(all.lip.samples$V1)
all.pairs <- as.data.frame(matrix(nrow = 0, ncol = 3, data = NA))
colnames(all.pairs) <- c("COLUMN","ROW","Percent.Ident")
for(i in 1:length(all.lip.samples)){
  column <- all.lip.samples[i]
  for(j in 1:(length(all.lip.samples)-i)){
    row.start <- i+1
    row.spp <- all.lip.samples[row.start:length(all.lip.samples)]
    row <- row.spp[j]
    percent.ident <- percent_identical[row,column]
    rowdat <- c(column, row, percent.ident)
    all.pairs <- rbind(all.pairs, rowdat)
  }
}
colnames(all.pairs) <- c("COLUMN","ROW","Percent.Ident")
all.pairs <- all.pairs[1:4465,]
all.pairs$Percent.Ident <- as.numeric(all.pairs$Percent.Ident)
```

Remove rows with same clade or same species pairs (should remove 197 and end up with 4268)
Get same clade (n = 162) and same spp (n = 35) dataframes as well
```{r}
all.same.clade <- read.csv("Identical_Sequences_files/same_clade_all.csv", header = FALSE)
colnames(all.same.clade) <- c("COLUMN","ROW")

same.clade.pairs <- read.csv("Identical_Sequences_files/same_clade_pairs.csv", header = FALSE)
colnames(same.clade.pairs) <- c("COLUMN","ROW")

same.spp.pairs <- read.csv("Identical_Sequences_files/same_spp_pairs.csv", header = FALSE)
colnames(same.spp.pairs) <- c("COLUMN","ROW")

nonclade.pairs <- anti_join(all.pairs, all.same.clade, by = c("COLUMN","ROW"))

clade.pairs <- inner_join(all.pairs, same.clade.pairs,  by = c("COLUMN","ROW"))

spp.pairs <- inner_join(all.pairs, same.spp.pairs,  by = c("COLUMN","ROW"))
```

Get results stats
```{r}
summary(spp.pairs$Percent.Ident) #min = 4.6 %; Mean = 25.2% +/- 10.7%; max = 47.9%; N = 35;
length(spp.pairs$Percent.Ident)
sd(spp.pairs$Percent.Ident)

summary(clade.pairs$Percent.Ident) #min = 8.9 %; Mean = 23.6% +/- 8.6%; max = 43.0%; N = 162;
length(clade.pairs$Percent.Ident)
sd(clade.pairs$Percent.Ident)

summary(nonclade.pairs$Percent.Ident) #min = 1.9 %; Mean = 14.4% +/- 4.1%; max = 36.6%; N = 4268;
length(nonclade.pairs$Percent.Ident)
sd(nonclade.pairs$Percent.Ident)

```

General Stats
Spp.pairs = min = 4.6 %; Mean = 25.2% +/- 10.7%; max = 47.9%; N = 35;
clade.pairs = min = 8.9 %; Mean = 23.6% +/- 8.6%; max = 43.0%; N = 162
nonclade.pairs = min = 1.9 %; Mean = 14.4% +/- 4.1%; max = 36.6%; N = 4268


Do stats to check if groups are different
```{r}
var.test(spp.pairs$Percent.Ident, clade.pairs$Percent.Ident) #var equal
var.test(spp.pairs$Percent.Ident, nonclade.pairs$Percent.Ident) #var diff
var.test(clade.pairs$Percent.Ident, nonclade.pairs$Percent.Ident) #var diff
```

```{r}
t.test(spp.pairs$Percent.Ident, clade.pairs$Percent.Ident, var.equal = TRUE)
t.test(spp.pairs$Percent.Ident, nonclade.pairs$Percent.Ident, var.equal = FALSE, alternative = "greater") #var diff
t.test(clade.pairs$Percent.Ident, nonclade.pairs$Percent.Ident, var.equal = FALSE, alternative = "greater") #var diff
```


Plot boxplots
```{r}
tiff(filename = "../../../../Desktop/Percent_ident_boxplot.tiff", width = 7, height = 5, units = "in", res = 300)
boxplot(spp.pairs$Percent.Ident, clade.pairs$Percent.Ident, nonclade.pairs$Percent.Ident, ylab = "Percent of Idenitcal Sequences",
        yaxt = "n", xaxt = "n", ylim = c(0, 0.51))
axis(2, at = seq(0,0.5, by = 0.1),labels = c("0%","10%","20%","30%","40%","50%"))
axis(1, at = c(1,2,3),labels = c("Same Species","Same Clade","Different Clades"))
axis(1, at = c(1,2,3), labels = c("N = 33", "N = 162", "N = 4268"), line = 1 , tick = FALSE)
text(x = 3, y = 0.5, "*", cex = 4)
dev.off()
```

## SNP read depth histograms

read in data
```{r}
all.depth <- read.csv("../12_bowtie/All_Dp_AD_combined.csv")
all.filt <- read.csv("../12_bowtie/All_Dp_AD_filtered.csv")
all.seq.filt <- all.filt %>% filter(MA_Freq > 0.01)
all.seq.filt2 <- all.filt %>% filter(MA_Freq > 0.05)
```

Read in and process PLEQUAD data from Calder's
```{r}
plequad.vcf <- vcf.parser("../12_bowtie/test5_calder/all_recoded_vcfs/PLEQUAD_UW151438_S22_results.vcf")
plequad.filt1 <- plequad.vcf[[9]]
plequad.filt2 <- plequad.filt1 %>% filter(MA_Freq >0.05)

acandu.vcf <- vcf.parser("../12_bowtie/test5_calder/all_recoded_vcfs/ACANADU_UW118829_S45_results.vcf")
acandu.filt1 <- acandu.vcf[[9]]
acandu.filt2 <- acandu.filt1 %>% filter(MA_Freq >0.05)

psemela.vcf <- vcf.parser("../12_bowtie/test5_calder/all_recoded_vcfs/PSEMELA_UW154235_S57_results.vcf")
psemela.filt1 <- psemela.vcf[[9]]
psemela.filt2 <- psemela.filt1 %>% filter(MA_Freq >0.05)

par(mfrow=c(2,3))
hist(all.seq.filt$MA_Freq, xlim = c(0,0.5), breaks = 10)
hist(psemela.filt2$MA_Freq, xlim = c(0,0.5), breaks = 10)
hist(acandu.filt2$MA_Freq, xlim = c(0,0.5), breaks = 10)
hist(plequad.filt2$MA_Freq, xlim = c(0,0.5), breaks = 10)

hist(all.seq.filt2$MA_Freq, xlim = c(0,0.5), breaks = 10)
```


Minor allele occurence stats
```{r}
mean(all.seq.filt2$MA_Freq) #0.24
sd(all.seq.filt2$MA_Freq) #0.13

sd(all.filt$Depth)

max(all.depth$Depth)
length(which(all.depth$Depth > 200))/length(all.depth$Depth)
```



Generate data for simulated allele frequency from expected distribution
```{r}
rtruncnorm(30602, a = 0, b = Inf, mean = 0, sd = 0.075)
bimod <- c(rnorm(30602, 0, 0.05), rnorm(30602, 0.5, 0.05))
bimod.trunc <- c(rtruncnorm(30602, a = -Inf, b = 0.5, mean = 0.5, sd = 0.01))
par(mfrow = c(2,1))
hist(bimod.trunc, xlim = c(0,0.5), breaks = 2, main = "", xlab = "")
axis(1, at = 0.25, labels = c("Rate of Occurence of Minor Allele"), line = 1.5, tick = FALSE)
axis(1, at = 0.25, labels = c("N = 61204 hypothetical SNPs"), line = 2.5, tick = FALSE)
hist(all.seq.filt$MA_Freq, main="", xlab = "", breaks = 20)
axis(1, at = 0.25, labels = c("Rate of Occurence of Minor Allele"), line = 1.5, tick = FALSE)
axis(1, at = 0.25, labels = c("N = 61204 filtered SNPs"), line = 2.5, tick = FALSE)

```


raw read depth hist
```{r}
tiff(filename = "../../../../Desktop/Read_Depth_hist.tiff", width = 7, height = 5, units = "in", res = 300)
hist(all.depth$Depth, main = "", xlab = "", xlim = c(0,325), ylim = c(0,50000))
abline(v=20, col = "red", lwd = 2)
axis(1, at = 150, labels = c("Read Depth per Locus"), line = 1.5, tick = FALSE)
axis(1, at = 150, labels = c("N = 105,844 filtered SNP Loci"), line = 2.5, tick = FALSE)
dev.off()
```

"minor" allele frequency for all snps after filtering
```{r}
tiff(filename = "../../../../Desktop/MA_freq_hist.tiff", width = 7, height = 8.5, units = "in", res = 300)
par(mfrow = c(2,1))
hist(plequad.filt2$MA_Freq, xlim = c(0,0.5), breaks = 10, main = "", xlab = "", freq = FALSE, ylab = "", yaxt = "n")
axis(1, at = 0.25, labels = c("Minor Allele Frequency at Heterozygous Sites"), line = 1.5, tick = FALSE)
axis(1, at = 0.25, labels = c("N = 762 filtered SNPs"), line = 2.5, tick = FALSE)
axis(2, at = seq(0,4, by =1), labels = c("0%", "10%", "20%", "30%","40%"))
axis(2, at = 2, labels = "Percentage of Sites", line = 1.5, tick = FALSE)
hist(all.seq.filt2$MA_Freq, xlim = c(0,0.5), breaks = 10, main="", xlab = "", freq = FALSE, ylab = "", yaxt = "n")
axis(1, at = 0.25, labels = c("Minor Allele Frequency at Heterozygous Sites"), line = 1.5, tick = FALSE)
axis(1, at = 0.25, labels = c("N = 28807 filtered SNPs"), line = 2.5, tick = FALSE)
axis(2, at = seq(0,4, by =1), labels = c("0%", "10%", "20%", "30%","40%"))
axis(2, at = 2, labels = "Percentage of Sites", line = 1.5, tick = FALSE)
dev.off()
```

read depth vs minor allele frequency test plots
```{r}
summary(gam(all.filt$Depth~s(all.filt$MA_Freq, bs = "cs"))) #r2 = 0.05
summary(lm(all.filt$Depth~all.filt$MA_Freq)) #r2 = 0.02
summary(lm(all.filt$Depth~poly(all.filt$MA_Freq,2))) #r2 = 0.03

ggplot(all.filt, aes(x=MA_Freq, y=Depth)) + geom_point()
ggplot(all.filt, aes(x=MA_Freq, y=Depth)) + geom_smooth(method = lm)
ggplot(all.filt, aes(x=MA_Freq, y=Depth)) + geom_smooth(formula = y ~ poly(x,2))
ggplot(all.filt, aes(x=MA_Freq, y=Depth)) + geom_smooth()

ggplot(all.filt, aes(x=MA_Freq, y=Depth)) + geom_point() +
  geom_smooth(method = lm, colour = "grey50") + 
  geom_smooth(colour = "orange") +
  geom_smooth(formula = y ~ poly(x,2), colour = "blue") 
  
```

Final Plot (quadratic only)
```{r}
tiff(filename = "../../../../Desktop/Depth_MAfreq.tiff", width = 7, height = 5, units = "in", res = 300)
ggplot(all.filt, aes(x=MA_Freq, y=Depth)) + 
  geom_point(colour = "grey80") + 
  theme_classic() +
  geom_smooth(formula = y ~ poly(x,2), colour = "black") + 
  labs(x = "Minor Allele Frequency at Heterozygous Sites", y = "Read Depth") +
  theme(axis.title.x = element_text(size=14, face="bold"),axis.title.y = element_text(size=14, face="bold"),
        axis.text=element_text(size=12, colour = "black")) +
  geom_text(x=0.48, y=320, label=expression(r^2 == 0.03))
dev.off()
```


## Making P-distance histogram
read in data
```{r}
p.dists <- read.csv("P_distances_bygene.csv", header = TRUE)
max(p.dists$avg_pdist)
mean(p.dists$avg_pdist)
sd(p.dists$avg_pdist)
```

make hist
```{r}
tiff(filename = "../../../../Desktop/pdist_hist.tiff", width = 7, height = 5, units = "in", res = 300)
hist(p.dists$avg_pdist, main = "", xlab = "Average p-distance for each of 2917 exons", xlim = c(0,0.2), ylim = c(0,900))
dev.off()
```

## Thinking about read depth
for illumina hiseq 3000 output is typically 300,000,000-400,000,000 reads per lane. Target 500 reads per gene per individual with 4434 loci. will go with lower estimate to be conservative.
Most genes recovered is 3753
Calder's average raw reads = 6.4 mill and trimmed reads = 6 mill
My data's average raw reads = 12 mill and trimmed reads = 11 mill - more reads but also more reads trimmed on average than Calder's

```{r}
11e6/4434 # should have read depth of over 2000 reads per indidual per locus -> so clearly a large proportion of reads are not mapping to genes (expected)
500*4434 # to get 500 reads per roughly 2 million reads per individual (reasonable goal)
300/25 # could reasonably expected roughtly 12 million reads per individual if only 25 in poopl
400/25 # up to 16 million reads per indidivual with upper limit of reads per lane from hiseq 3000.

300/11

```



## Making Pretty Trees
Filenames_bootstrap.whatever is the file with all raxml trees present but no branch lengths

```{r}
run2 <- read.tree(file = "../4_Run_2/9_RAxML_orn_rm_bl/RAxML_bootstrap.9_orn_bl")
run2.consensus <- consensus(run2, p = 0.5, check.labels = TRUE, rooted = FALSE)
run2.consensus.bl <- consensus.edges(run2)
plotTree(run2.consensus)
plotTree(run2.consensus.bl)
write.tree(run2.consensus.bl, file = "TreeFigures/Run2_consensus.tree", digits = 20)
```

```{r}
run6 <- read.tree(file = "../9_Run_6/28_RAxML_r6_bl/RAxML_bootstrap.28_r6_bl")
run6.consensus <- consensus(run6, p = 0.5, check.labels = TRUE, rooted = FALSE)
run6.consensus.bl <- consensus.edges(run6)
plotTree(run6.consensus)
plotTree(run6.consensus.bl)
write.tree(run6.consensus.bl, file = "TreeFigures/Run6_consensus.tree", digits = 20)
```

```{r}
rungene <- read.tree(file = "../10_Gene_Trees/32_RAxML_full_bl/RAxML_bootstrap.32_full_bl")
rungene.consensus <- consensus(rungene, p = 0.5, check.labels = TRUE, rooted = FALSE)
rungene.consensus.bl <- consensus.edges(rungene)
plotTree(rungene.consensus)
plotTree(rungene.consensus.bl)
write.tree(rungene.consensus.bl, file = "TreeFigures/Rungene_consensus.tree", digits = 20)
```
```{r}
run.da <- read.tree(file = "../0_da_test/Da_Tree/RAxML_bootstrap.DA_tree2_bl")
run.da.consensus.bl <- consensus.edges(run.da)
plotTree(run.da.consensus.bl)
write.tree(run.da.consensus.bl, file = "TreeFigures/Runda_consensus.tree", digits = 20)
```