---
title: "List of Identical Sequences"
author: "Jennifer Gardner"
date: "10/28/2020"
output: pdf_document
---
load dependencies
```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(phangorn)
library(ape)
```



### Goals:
* Look at each alignment file for each gene and flag taxa that have identical sequences
  * p-distance = 0
* Return list of taxa with identical sequences
  * Flag those that are different species (not concerned if same species have identical sequences)
* Make a taxa-by-taxa table with sums of number of genes where two taxa have identical sequences
*Make a second table with list of genes and number of identical sequences within each gene
  * This will help identify contamination vs. highly conserved loci

#### Step 1: Figure out how to do it for one locus

Use files from 4_Run_2/7_nf_aligned_orn/. Run 2 removed Oreochromis at the pick_taxa step and aligned sequences without it. Run 2 included all taxa and all genes with at least 49 taxa (50%) present. 
```{r}
gene <- read.FASTA("../4_Run_2/7_nf_aligned_orn/Danio_rerio.1.10132455.10132285.fas")
gene.dist <- as.data.frame(dist.dna(gene, as.matrix = TRUE))
columns <- colnames(gene.dist)
rows <- rownames(gene.dist)
gene.dist1 <- as.matrix(gene.dist)
colnames(gene.dist1) <- NULL
rownames(gene.dist1) <- NULL
taxa <- length(gene.dist)
pzeros <- which(gene.dist1==0, arr.ind = TRUE)
pzeros1 <- pzeros[which(pzeros[,1]!=pzeros[,2]),]
for (i in 1:length(pzeros)){
  index <- pzeros[i]
  column <- ifelse(round(index/taxa)>index/taxa,round(index/taxa),round(index/taxa)+1)
}
```

#### Step 2: Figure out how to loop over directory

```{r}
filenames <- list.files(path = "../4_Run_2/7_nf_aligned_orn/")
head(filenames)
```

